Mathlib/RingTheory/PowerSeries/WeierstrassPreparation.lean

```diff
@@ -0,0 +1,717 @@
+  the formal power series `g` satisfies `g = f * h`.
+
+- `PowerSeries.IsWeierstrassFactorization`: version of `PowerSeries.IsWeierstrassFactorizationAt`
+  for local rings with respect to its maximal ideal.
+
+## Main results
+
+- `PowerSeries.exists_isWeierstrassDivision`: **Weierstrass division**: let `f`, `g` be power
+  series over a complete local ring, such that the image of `g` in the residue field is not zero.
+  Let `n` be the order of the image of `g` in the residue field. Then there exists a power series
+  `q` and a polynomial `r` of degree `< n`, such that `f = g * q + r`.
+
+- `PowerSeries.IsWeierstrassDivision.elim`: The `q` and `r` in the Weierstrass division is unique.
+
+- `PowerSeries.exists_isWeierstrassFactorization`: **Weierstrass preparation theorem**: let `g` be
+  a power series over a complete local ring, such that the image of `g` in the residue field is
+  not zero. Then there exists a distinguished polynomial `f` and a power series `h`
+  which is a unit, such that `g = f * h`.
+
+- `PowerSeries.IsWeierstrassFactorization.elim`: The `f` and `h` in Werierstrass preparation
+  theorem is unique.
+
+-/
+
+open scoped Polynomial PowerSeries
+
+/-!
+
+## Results should be added to mathlib
+
+-/
+
+theorem ENat.lift_eq_toNat {x : ℕ∞} (hx : x < ⊤) : x.lift hx = x.toNat := by
+  rcases x with ⟨⟩ | x
+  · contradiction
+  · rfl
+
+-- unused
+theorem PowerSeries.eq_of_X_pow_mul_eq_X_pow
+    {A : Type*} [CommRing A] [Nontrivial A] (m n : ℕ) (u : A⟦X⟧) (hu : IsUnit u)
+    (h : X ^ m * u = X ^ n) : m = n := by
+  rw [isUnit_iff_constantCoeff] at hu
+  replace h := congr(coeff A m $h)
+  rw [coeff_mul, ← Finset.sum_subset (s₁ := {(m, 0)}) (by simp) (fun p hp hnmem ↦ by
+    have hp1 : p.1 ≠ m := by
+      contrapose! hnmem
+      rwa [Finset.mem_singleton, Finset.antidiagonal_congr hp (by simp)]
+    rw [coeff_X_pow, if_neg hp1, zero_mul]), Finset.sum_singleton, coeff_X_pow_self,
+    coeff_zero_eq_constantCoeff, one_mul, coeff_X_pow] at h
+  split_ifs at h with hmn
+  · exact hmn
+  · exact False.elim (hu.ne_zero h)
+
+-- unused
+theorem Polynomial.IsDistinguishedAt.natDegree_eq_of_associated_powerSeries
+    {A : Type*} [CommRing A] {I : Ideal A} {f g : A[X]}
+    (hf : f.IsDistinguishedAt I) (hg : g.IsDistinguishedAt I) (hI : I ≠ ⊤)
+    (hfg : Associated (f : A⟦X⟧) (g : A⟦X⟧)) : f.natDegree = g.natDegree := by
+  rcases subsingleton_or_nontrivial (A ⧸ I) with h | _
+  · apply (Submodule.Quotient.subsingleton_iff _).1 at h
+    exact False.elim <| hI <| by ext; simp [h]
+  obtain ⟨u, hu⟩ := hfg
+  have hu' := congr(PowerSeries.map (Ideal.Quotient.mk I) $hu)
+  simp_rw [map_mul, ← Polynomial.polynomial_map_coe, hf.map_eq_X_pow, hg.map_eq_X_pow,
+    Polynomial.coe_pow, Polynomial.coe_X] at hu'
+  change _ * (RingHom.toMonoidHom _) _ = _ at hu'
+  rw [← Units.coe_map] at hu'
+  exact PowerSeries.eq_of_X_pow_mul_eq_X_pow _ _ _ (Units.isUnit _) hu'
+
+namespace PowerSeries
+
+section
```

Comment:
This whole section should be moved somewhere else.

Reply:
I haven't find a good place to put them into. Seems that there are no file in `PowerSeries` directory imports both `PowerSeries.coeff_mul` and `Ideal.mul_mem_mul`.

Reply:
Done. They are split to a new file.